### PR TITLE
More typing improvements to `@endo/eventual-send` and `@endo/captp`

### PR DESCRIPTION
**Pull request by michaelfig**
These improve the typing of `E` and `Trap` to be more accurate.

Also, introduce `FarRef<T>` as an "I can handle near, far objects and promises for them" type.

### CODE REVIEW

File: packages/eventual-send/src/index.d.ts
Comment (mhofman) [inline packages/eventual-send/src/index.d.ts:6]: I think we can get rid of this and use the native `PropertyKey` instead

File: packages/captp/jsconfig.test.json
Comment (mhofman) [inline packages/captp/jsconfig.test.json:3]: `"src/index.d.ts"` should be covered by `"src/**/*.ts"`

File: packages/eventual-send/jsconfig.test.json
Comment (mhofman) [inline packages/eventual-send/jsconfig.test.json:3]: `"src/index.d.ts"` should be covered by `"src/**/*.ts"`

**Review by mhofman — COMMENTED** [review]
Looks great!

Let's include the `"skipLibCheck": false` in at least the `eventual-send` `jsconfig.json` as I've confirmed locally that doesn't cause issues. (See https://github.com/Agoric/agoric-sdk/issues/4735)

`const E` should be `declare`d (even though that's implicit from being a `d.ts` file):
```ts
export declare const E: EProxy;
```

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:1]: will this file not be checked without this?

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:6]: I don't follow why we need `Callable` instead of `Function`. ??

If we do, should `FunctionOnly` and `RemoteFunction` also change?

File: packages/eventual-send/src/index.d.ts
Comment (michaelfig) [reply packages/eventual-send/src/index.d.ts:1]: > will this file not be checked without this?

Your implication is right: it's unnecessary.

@mhofman and I were experimenting with trying to get the file checked by `tsc` when it really wasn't.  Turns out `"skipLibCheck": false` was the missing requirement.  IOU an issue.

File: packages/eventual-send/src/index.d.ts
Comment (michaelfig) [reply packages/eventual-send/src/index.d.ts:6]: The problem is that `typeof x extends Function` does not mean that `typeof x === 'function'`, but `typeof x extends Callable` does.

We rely on `typeof x === 'function'` deep in the bowels of the `E` proxy maker.

File: packages/eventual-send/src/index.d.ts
Comment (mhofman) [reply packages/eventual-send/src/index.d.ts:1]: Issue: https://github.com/Agoric/agoric-sdk/issues/4735

File: packages/eventual-send/src/index.d.ts
Comment (mhofman) [reply packages/eventual-send/src/index.d.ts:6]: I understood @turadg's question as whether `FunctionOnly` or `RemoteFunction` are the right names, or if it should be `CallableOnly` and `RemoteCallable`. I think it's fine to keep `Callable` internally to help with inference, and `Function` as the external naming, as people are more familiar with that terminology.

**Comment by turadg** [comment]
Another "type improvement" we should include before landing this is code docs for how these improved types work. @erights provides specific questions in https://github.com/endojs/endo/pull/1095#pullrequestreview-903720221

I'd be happy to take a pass at such documentation. @michaelfig do you want to handle that or shall I push a commit?

**Comment by michaelfig** [comment]
> I'd be happy to take a pass at such documentation.

I'd appreciate that.  Feel free to rebase, and ping me when you'd like to pass the baton back.

**Comment by erights** [comment]
> > I'd be happy to take a pass at such documentation.
> 
> I'd appreciate that. Feel free to rebase, and ping me when you'd like to pass the baton back.

@turadg just FYI, I found @michaelfig responses to my comments on #1095 helpful. When you do have that PR, please include me as a reviewer. Thanks!

**Review by erights — CHANGES_REQUESTED** [review]
@michaelfig and I had a long conversation that suggested several changes. Please ask me to review again after those are done. Thanks!

**Comment by michaelfig** [comment]
> @michaelfig and I had a long conversation that suggested several changes. Please ask me to review again after those are done. Thanks!

They're done!

**Comment by turadg** [comment]
I got some errors when I tried this in my open PRs. I think they're just errors in master though:

<details>
 <summary>agoric master with `yarn link "@endo/eventual-send"` </summary>

```
src/econ-behaviors.js:146:34 - error TS2349: This expression is not callable.
  Type 'EMethods<Required<VatLoader<PriceAuthorityVat>>>' has no call signatures.

146   const vats = { priceAuthority: E(loadVat)('priceAuthority') };
                                     ~~~~~~~~~~

src/econ-behaviors.js:399:34 - error TS2349: This expression is not callable.
  Type 'EMethods<Required<VatLoader<DistributeFeesVat>>>' has no call signatures.

399   const vats = { distributeFees: E(loadVat)('distributeFees') };
                                     ~~~~~~~~~~

../vats/src/distributeFees.js:46:53 - error TS2345: Argument of type 'ERef<Payment<AssetKind>>' is not assignable to parameter of type 'Payment<AssetKind>'.
  Property 'getAllegedBrand' is missing in type 'PromiseLike<Payment<AssetKind>>' but required in type 'Payment<AssetKind>'.

46         .then(payment => E(feeDepositFacet).receive(payment)),
                                                       ~~~~~~~

  ../ERTP/src/types.js:373:4
    373  * @property {() => Brand<K>} getAllegedBrand
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    374  * Get the allegedBrand, indicating the type of digital asset this
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
    377  * with suspicion and verified elsewhere.
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    378  */
        ~
    'getAllegedBrand' is declared here.

../vats/src/vat-bank.js:299:11 - error TS2322: Type 'Promise<ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>>' is not assignable to type 'ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>'.
  Type 'Promise<ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>>' is not assignable to type 'PromiseLike<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>'.
    Types of property 'then' are incompatible.
      Type '<TResult1 = ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>, TResult2 = never>(onfulfilled?: ((value: ERef<...> | ERef<...>) => TResult1 | PromiseLike<...>) | ... 1 more ... | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | ... 1...' is not assignable to type '<TResult1 = { receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }, TResult2 = never>(onfulfilled?: ((value: { receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<...>; }) => TResult1 | PromiseLike<...>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<.....'.
        Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
          Types of parameters 'value' and 'value' are incompatible.
            Type 'ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>' is not assignable to type '{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }'.
              Property 'receive' is missing in type 'PromiseLike<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>' but required in type '{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }'.

299           return E(vp).getDepositFacet();
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  ../ERTP/src/types.js:307:4
    307  * @property {DepositFacetReceive} receive
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    308  * Deposit all the contents of payment into the purse that made this facet,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
    312  * If payment is a promise, throw an error.
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    313  */
        ~
    'receive' is declared here.

```
</details>


<details>
 <summary> https://github.com/Agoric/agoric-sdk/pull/4831 </summary>

```
src/econ-behaviors.js:146:34 - error TS2349: This expression is not callable.
  Type 'EMethods<Required<VatLoader<PriceAuthorityVat>>>' has no call signatures.

146   const vats = { priceAuthority: E(loadVat)('priceAuthority') };
                                     ~~~~~~~~~~

src/econ-behaviors.js:399:34 - error TS2349: This expression is not callable.
  Type 'EMethods<Required<VatLoader<DistributeFeesVat>>>' has no call signatures.

399   const vats = { distributeFees: E(loadVat)('distributeFees') };
                                     ~~~~~~~~~~

../vats/src/distributeFees.js:46:53 - error TS2345: Argument of type 'ERef<Payment<AssetKind>>' is not assignable to parameter of type 'Payment<AssetKind>'.
  Property 'getAllegedBrand' is missing in type 'PromiseLike<Payment<AssetKind>>' but required in type 'Payment<AssetKind>'.

46         .then(payment => E(feeDepositFacet).receive(payment)),
                                                       ~~~~~~~

  ../ERTP/src/types.js:373:4
    373  * @property {() => Brand<K>} getAllegedBrand
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    374  * Get the allegedBrand, indicating the type of digital asset this
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
    377  * with suspicion and verified elsewhere.
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    378  */
        ~
    'getAllegedBrand' is declared here.

../vats/src/vat-bank.js:299:11 - error TS2322: Type 'Promise<ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>>' is not assignable to type 'ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>'.
  Type 'Promise<ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>>' is not assignable to type 'PromiseLike<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>'.
    Types of property 'then' are incompatible.
      Type '<TResult1 = ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>, TResult2 = never>(onfulfilled?: ((value: ERef<...> | ERef<...>) => TResult1 | PromiseLike<...>) | ... 1 more ... | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | ... 1...' is not assignable to type '<TResult1 = { receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }, TResult2 = never>(onfulfilled?: ((value: { receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<...>; }) => TResult1 | PromiseLike<...>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<.....'.
        Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
          Types of parameters 'value' and 'value' are incompatible.
            Type 'ERef<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }> | ERef<DepositFacet>' is not assignable to type '{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }'.
              Property 'receive' is missing in type 'PromiseLike<{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }>' but required in type '{ receive: (args_0: Payment<AssetKind>, args_1?: any) => ERef<Amount<AssetKind>>; }'.

299           return E(vp).getDepositFacet();
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  ../ERTP/src/types.js:307:4
    307  * @property {DepositFacetReceive} receive
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    308  * Deposit all the contents of payment into the purse that made this facet,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
    312  * If payment is a promise, throw an error.
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    313  */
        ~
    'receive' is declared here.

```
</details>


So it looks like it doesn't yet handle the case of an arg that's an `ERef` for a callable. The function overload is matching that to the `object of callables` case.

Did you try this with master? I wonder if I'm doing something differently. Given how critical this dep is to the functioning of agoric-sdk and that we don't have types integration tests, maybe https://github.com/endojs/endo/issues/1118 is something to prioritize.

**Comment by michaelfig** [comment]
> I got some errors when I tried this in my open PRs. I think they're just errors in master though:

Thanks for doing this.  I've also tested against master now, and made tweaks to this PR to support it.

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:205]: Is `Awaited` necessary here? The condition on 203 implies that `T[P]` is not PromiseLike. 

Though the fact that `Awaited` works here make me think we don't need the conditional:
```
readonly [P in keyof T]: T[P];
```
And in that case why do we need `EGetters` ? Perhaps it needs a `never` case for members that aren't getters.

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:32]: conflicting comment styles. I think the `//` suffices, it's not user documentation.

if you want it to be then let's use `@see {@link …`

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:92]: This name is confusing given the comment above,
> `RemoteCallable<T>` means to return a record type consisting only of properties that are functions.

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:86]: What's this `T2` business? :) 
```suggestion
 * `RemoteCallable<T>` means to return a record type consisting only of properties that are functions.
```

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:176]: 👍 moving this down

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:185]: these comments lost the newline before them. they apply to the line below each comment.

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:192]: this is more of a transform. "ensure" makes me think there'd be an error for anything that doesn't satisfy the condition

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:194]: Also what about members that aren't functions, how should those be treated? Consider a type constraint here

**Review by turadg — CHANGES_REQUESTED** [review]
Looking good. I had some questions about the use of `Awaited` and a blocking request. (To be resolved by using function overloading or a good argument why not :)

**Comment by turadg** [comment]
Requesting one more change: that the type checks here use https://github.com/SamVerschueren/tsd/ added in #1127

**Comment by michaelfig** [comment]
> Requesting one more change: that the type checks here use https://github.com/SamVerschueren/tsd/ added in #1127

If I want to maintain dynamic checks, then I should not be prevented from doing that.

I agree that static-only checks can be useful, but if they are just a subset of the checks I'm already doing in my dynamic tests, that's not useful.

So I'll punt back to you: I want to keep `eventual-send/test/test-types.js`.  Is there anything that you think static checks would add that would warrant having them in addition to this mixed dynamic/static test?

**Comment by turadg** [comment]
👍 to your statements.

> Is there anything that you think static checks would add that would warrant having them in addition to this mixed dynamic/static test?

Yes, I think the specificity like https://github.com/endojs/endo/blob/47d6bc814e68dea0a28864542923f4f186676ce2/packages/eventual-send/src/index.test-d.ts#L31 is worth testing/documenting. Feel free to grab  47d6bc814.

File: packages/eventual-send/src/index.d.ts
Comment (michaelfig) [reply packages/eventual-send/src/index.d.ts:205]: I need this construct for union types like `Foo | PromiseLike<Bar>` which combined are not `PromiseLike`.  Then the result is `Promise<Foo | Bar>`.  But the first `extends` is necessary so that I don't strip away template parameters by repackaging something that is already `PromiseLike`.

File: packages/eventual-send/src/index.d.ts
Comment (michaelfig) [reply packages/eventual-send/src/index.d.ts:237]: I don't know how to write the function overload.  Can you provide an example that properly tests `RemoteFunctions<T> extends Callable` and isn't overly verbose?

**Comment by michaelfig** [comment]
> Feel free to grab [47d6bc8](https://github.com/endojs/endo/commit/47d6bc814e68dea0a28864542923f4f186676ce2).

I think I'll leave this for a future PR, but thanks for getting that started.

> Though I retract gating my approval on using tsd.

Okay, thanks.

**Comment by michaelfig** [comment]
@erights please dismiss your request for changes, and rereview as desired.  This is ready to land.

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [reply packages/eventual-send/src/index.d.ts:237]:
```suggestion
  <T extends Callable>(x: T): ECallable<T> & EMethods<Required<T>>;
  <T>(x: T): EMethods<Required<T>>;
```

[playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDCBDANljAjLVAXjgAoA6SjKAcwGcAuODAO0QG0BdASjiID5mbANwBYAFASAlixjAoAMwwBjVAFEAClAghEcAN4S4cADwAVOKDksAJnXTZcBYP1IgmZ7kwDkyx-kJvMXFjc1d3OE8fAFcWPxwA4CCJAF8JCQB6DLgAARg6AFpQFGUYIqhtKAllCBY6eDUmTW1dPjgWaJxgiTVSb3qob25hOCy4b1j4p0Ce0lJeAXbOrF4x339nbwkgA)

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:4]: 👏

File: packages/eventual-send/src/index.d.ts
Comment (turadg) [inline packages/eventual-send/src/index.d.ts:223]: this could also have a function overload.

if you don't want to include in this PR I can follow up

**Review by turadg — APPROVED** [review]
Approving because this is a definite improvement and it looks like most concerns I raised were addressed. The function overloading could wait for another PR.

I trust that you've checked with Agoric SDK master again after latest commits.

File: packages/eventual-send/src/index.d.ts
Comment (michaelfig) [reply packages/eventual-send/src/index.d.ts:223]: > if you don't want to include in this PR I can follow up

Yes, please.  Let's chat about it before you spend much time; I think the proposals you made don't deal with `RemoteFunctions<T>` correctly.

**Pull request merged**
